Mail/Proxy/Implementation/MessageActionHandler.swift

```diff
@@ -0,0 +1,128 @@
+/*
+ Infomaniak Mail - iOS App
+ Copyright (C) 2022 Infomaniak Network SA
+
+ This program is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import Foundation
+import InfomaniakCoreUI
+import InfomaniakDI
+import MailCore
+import RealmSwift
+import UIKit
+import UserNotifications
+
+public struct MessageActionHandler: MessageActionHandlable {
+    private enum ErrorDomain: Error {
+        case messageNotFoundInDatabase
+    }
+
+    private enum ActionNames {
+        static let archive = "archiveClicked"
+        static let delete = "deleteClicked"
+        static let reply = "reply"
+        static let open = "open"
+
+        static let archiveExecuted = "archiveExecuted"
+        static let deleteExecuted = "deleteExecuted"
+    }
+
+    @LazyInjectService private var accountManager: AccountManager
+    @LazyInjectService private var matomo: MatomoUtils
+
+    func handleTapOnNotification(messageUid: String, mailbox: Mailbox, mailboxManager: MailboxManager) async {
+        matomo.track(eventWithCategory: .notificationAction, name: ActionNames.open)
+
+        // Switch account if needed
+        switchAccountIfNeeded(mailbox: mailbox, mailboxManager: mailboxManager)
+
+        NotificationCenter.default.post(name: .onUserTappedNotification,
+                                        object: NotificationTappedPayload(messageId: messageUid))
+    }
+
+    func handleReplyOnNotification(messageUid: String, mailbox: Mailbox, mailboxManager: MailboxManager) {
+        matomo.track(eventWithCategory: .notificationAction, name: ActionNames.reply)
+
+        // Switch account if needed
+        switchAccountIfNeeded(mailbox: mailbox, mailboxManager: mailboxManager)
+
+        NotificationCenter.default.post(name: .onUserTappedReplyToNotification,
+                                        object: NotificationTappedPayload(messageId: messageUid))
+    }
+
+    func handleArchiveOnNotification(messageUid: String, mailbox: Mailbox, mailboxManager: MailboxManager) async throws {
+        matomo.track(eventWithCategory: .notificationAction, name: ActionNames.archive)
```

Comment:
I need to wrap this into a `UIApplication.shared.beginBackgroundTask` for the update to terminate updating corectly mail folders and unread counts

Reply:
Packaged into `UIApplicationBackgroundTaskHelper`, moving it to CoreUI.

Reply:
using ApplicationBackgroundTaskTracker where needed.